File: parkeye-backend/app/schemas/lot.py
"""Lot schemas."""
from uuid import UUID
from pydantic import BaseModel, Field


class LotResponse(BaseModel):
    """Single lot with current occupancy and status."""

    id: UUID
    name: str
    capacity: int
    permit_types: list[str]
    lat: float
    lon: float
    is_deck: bool
    floors: int | None
    status: str
    status_reason: str | None
    occupancy_pct: float
    color: str


class LotDetailResponse(LotResponse):
    """Single lot with upcoming events affecting it."""

    upcoming_events: list[dict] = Field(default_factory=list)


class LotListResponse(BaseModel):
    """List of lots."""

    lots: list[LotResponse]


class LotHistoryPoint(BaseModel):
    """Single point in occupancy history."""

    hour_of_day: int
    day_of_week: int
    occupancy_pct: float
    color: str


class LotHistoryResponse(BaseModel):
    """Hourly occupancy for past 7 days."""

    lot_id: UUID
    history: list[dict]


class FloorOccupancy(BaseModel):
    """Per-floor occupancy for parking decks."""

    floor: int
    occupancy_pct: float
    color: str


class LotFloorsResponse(BaseModel):
    """Per-floor occupancy breakdown."""

    lot_id: UUID
    floors: list[FloorOccupancy]

File: parkeye_frontend/parkeye/Models/Lot.swift
import Foundation

struct Lot: Decodable, Identifiable {
    let id: UUID
    let name: String
    var occupancyPct: Double
    var color: String
    let capacity: Int
    let permitTypes: [String]?
    let lat: Double
    let lon: Double
    let isDeck: Bool
    let floors: Int?
    let status: String?
    let statusUntil: Date?
    let statusReason: String?

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(UUID.self, forKey: .id)
        name = try c.decode(String.self, forKey: .name)
        let pct = try c.decode(Double.self, forKey: .occupancyPct)
        occupancyPct = pct * 100
        color = hexFromSemanticColor(try c.decode(String.self, forKey: .color))
        capacity = try c.decode(Int.self, forKey: .capacity)
        permitTypes = try c.decodeIfPresent([String].self, forKey: .permitTypes)
        lat = try c.decode(Double.self, forKey: .lat)
        lon = try c.decode(Double.self, forKey: .lon)
        isDeck = try c.decode(Bool.self, forKey: .isDeck)
        floors = try c.decodeIfPresent(Int.self, forKey: .floors)
        status = try c.decodeIfPresent(String.self, forKey: .status)
        statusUntil = try c.decodeIfPresent(Date.self, forKey: .statusUntil)
        statusReason = try c.decodeIfPresent(String.self, forKey: .statusReason)
    }

    private enum CodingKeys: CodingKey {
        case id, name, occupancyPct, color, capacity, permitTypes
        case lat, lon, isDeck, floors, status, statusUntil, statusReason
    }
}
